timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java

```diff
@@ -0,0 +1,117 @@
+
+    private static final Logger log = LoggerFactory.getLogger(AvailableTimestamps.class);
+
+    private final LastReturnedTimestamp lastReturnedTimestamp;
+    private final PersistentUpperLimit upperLimit;
+    private volatile Throwable previousAllocationFailure;
+
+    public AvailableTimestamps(LastReturnedTimestamp lastReturnedTimestamp, PersistentUpperLimit upperLimit) {
+        this.lastReturnedTimestamp = lastReturnedTimestamp;
+        this.upperLimit = upperLimit;
+    }
+
+    public synchronized TimestampRange handOut(long numberToHandOut) {
+        checkArgument(
+                numberToHandOut <= MAX_TIMESTAMPS_TO_HAND_OUT,
+                "Can only hand out %s timestamps at a time, but %s were requested",
+                MAX_TIMESTAMPS_TO_HAND_OUT, numberToHandOut);
+
+        return handOutTimestamp(lastHandedOut() + numberToHandOut);
+    }
+
+    private synchronized TimestampRange handOutTimestamp(long targetTimestamp) {
+        checkArgument(
+                targetTimestamp > lastHandedOut(),
+                "Could not hand out timestamp '%s' as it was earlier than the last handed out timestamp: %s",
+                targetTimestamp, lastHandedOut());
+
+        TimestampRange rangeToHandOut = TimestampRange.createInclusiveRange(lastReturnedTimestamp.get() + 1, targetTimestamp);
+
+        allocateEnoughTimestampsToHandOut(targetTimestamp);
+        lastReturnedTimestamp.increaseToAtLeast(targetTimestamp);
+
+        return rangeToHandOut;
+    }
+
+    private long lastHandedOut() {
+        return lastReturnedTimestamp.get();
+    }
+
+    public synchronized void refreshBuffer() {
+        long buffer = upperLimit.get() - lastReturnedTimestamp.get();
+
+        if(buffer < MINIMUM_BUFFER || !upperLimit.hasIncreasedWithin(1, MINUTES)) {
+            allocateEnoughTimestampsToHandOut(lastHandedOut() + ALLOCATION_BUFFER_SIZE);
+        }
+    }
+
+    public synchronized void fastForwardTo(long newMinimum) {
+        lastReturnedTimestamp.increaseToAtLeast(newMinimum);
+        upperLimit.increaseToAtLeast(newMinimum + ALLOCATION_BUFFER_SIZE);
+    }
+
+    private void allocateEnoughTimestampsToHandOut(long timestamp) {
+        try {
+            upperLimit.increaseToAtLeast(timestamp);
+        } catch(Throwable e) {
+            handleAllocationFailure(e);
+        }
+    }
+
+    private synchronized void handleAllocationFailure(Throwable newFailure) {
+        Throwable oldFailure = previousAllocationFailure;
+        previousAllocationFailure = newFailure;
+
+        if (newFailure instanceof MultipleRunningTimestampServiceError) {
+            throw new ServiceNotAvailableException("This server is no longer valid because another is running.", newFailure);
+        }
+
+        if (newFailure != null) {
+            throw new RuntimeException("failed to allocate more timestamps", newFailure);
+        }
+
+        if (Thread.interrupted()) {
+            Thread.currentThread().interrupt();
```

Comment:
`Thread.interrupted()` delegates to `Thread.currentThread().isInterrupted(true)` which clears the interrupt flag. My nit here is that we're clearing then resetting the interrupt flag, which seems sketchy as what we really want is to just check `if (Thread.currentThread().isInterrupted())`. I realize this is based on existing code, I was just hoping to clean it up while we touched it :)